RetailApp/RetailApp.BAL/Providers/BaseProvider.cs

```diff
@@ -1,16 +1,18 @@
 ﻿using System;
 using System.Linq;
+using AutoMapper;
 using RetailApp.Data.Repository.Interfaces;
 
 namespace RetailApp.BAL.Providers
 {
     public abstract class BaseProvider<T> where T : class
     {
-        protected readonly IDbContextRepository<T> _repository;
+        protected readonly IMapper _mapper;
+        protected IDbContextRepository<T> _repository;
 
-        public BaseProvider(IDbContextRepository<T> repository)
+        public BaseProvider(IMapper mapper)
```

Comment:
Should also accept DbContextTypes as parameter. Create a private method or some other class(factory) to create new instance of the DbContext based on the DbContextTypes enum. After you have the dbContext instance, you could create a new instance of Repository and save it as a private variable